### PR TITLE
Fix broken links in CodeQL documentation

### PR DESCRIPTION
**Pull request by shati-patel**
### Overview

There are a number of broken links on codeql.github.com, some of which are probably due to the slightly different build process.

I ran the Sphinx linkchecker locally in "nitpicky" mode (`sphinx-build -n -b linkcheck <input-dir> <output-dir>`) to get a full list of broken and redirected links. Some of them were false positives, but most were helpful 🙂 

One of the key problems is that we can't use relative links to link to specific _headers_ in other articles. We either have to:
- Add an explicit reference (`.. example:`) and then link to it with a `:ref:` directive (e.g. ``` :ref:`Link text <example>` ```).
- Use the full URL (e.g. ``` `Link text <https://codeql.github.com/docs/article-name/#example>`__ ```). 

I've used the latter in most places, since it was easier than adding explicit reference keys everywhere. Also, if we eventually transfer the docs to another format (like markdown), we'd lose those `:ref:`s anyway, so it didn't seem worth the hassle.

### Changes

First commit (https://github.com/github/codeql/pull/4925/commits/3da66b7fd9bc5c27f241b35d5388883e77623ed0) fixes the genuine broken links, the second (https://github.com/github/codeql/pull/4925/commits/cdcb4a959971f3604917419d3320314e2fda2a92) fixes links that are redirected (not quite as important, but still nice to have!)

### CODE REVIEW

**Review by felicitymay — DISMISSED** [review]
Thanks for all these boring fixes. Great to have these links corrected. One small bug? but otherwise not much else to say here 😄

File: docs/codeql/ql-language-reference/about-the-ql-language.rst [hunk 1]
```diff
@@ -64,9 +64,9 @@ Here are a few prominent conceptual and functional differences between general p
 Further reading
 ---------------
 
-`Academic references <https://help.semmle.com/publications.html>`__ also provide an overview of QL and its semantics. Other useful references on database query languages and Datalog:
+`Academic references <https://codeql.github.com/publications/>`__ also provide an overview of QL and its semantics. Other useful references on database query languages and Datalog:
 
--  `Database theory: Query languages <http://www.lsv.ens-cachan.fr/~segoufin/Papers/Mypapers/DB-chapter.pdf>`__
--  `Logic Programming and Databases book - Amazon page <http://www.amazon.co.uk/Programming-Databases-Surveys-Computer-Science/dp/3642839541>`__
+-  `Database theory: Query languages <http://www.lsv.fr/~segoufin/Papers/Mypapers/DB-chapter.pdf>`__
+-  `Logic Programming and Databases book - Amazon page <https://www.amazon.co.uk/Programming-Databases-Surveys-Computer-Science/dp/3642839541>`__
```
[the remark this answers is not in the document]
Comment (shati-patel) [reply docs/codeql/ql-language-reference/about-the-ql-language.rst:70]: Good point! I'll use the Semantic Scholar link 👍🏽 

EDIT: The Semantic Scholar page links to Springer, so I might as well use that instead 😄

File: docs/codeql/ql-language-reference/expressions.rst [hunk 2]
```diff
@@ -63,7 +63,7 @@ You can express certain values directly in QL, such as numbers, booleans, and st
     "hello"
     "They said, \"Please escape quotation marks!\""
   
-  See `String literals <ql-language-specification#string-literals-string>`_ 
+  See `String literals <https://codeql.github.com/docs/ql-language-reference/ql-language-specification/#string-literals-string>`
```
[the remark this answers is not in the document]
Comment (shati-patel) [reply docs/codeql/ql-language-reference/expressions.rst]: Oooooooops 🙈 Fixed now!

**Review by felicitymay — APPROVED** [review]
All LGTM now :shipit:

**Pull request merged**